subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/JarContentHasher.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.changedetection.state;
+
+import com.google.common.hash.HashCode;
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.io.input.CloseShieldInputStream;
+import org.gradle.api.UncheckedIOException;
+import org.gradle.api.internal.cache.StringInterner;
+import org.gradle.util.DeprecationLogger;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipException;
+import java.util.zip.ZipInputStream;
+
+/**
+ * Hashes the contents of a jar file with the given {@link ContentHasher}.
+ */
+public class JarContentHasher implements ContentHasher {
+    private final ContentHasher contentHasher;
+    private final StringInterner stringInterner;
+
+    public JarContentHasher(ContentHasher contentHasher, StringInterner stringInterner) {
+        this.contentHasher = contentHasher;
+        this.stringInterner = stringInterner;
+    }
+
+    public HashCode hash(RegularFileSnapshot jarFile) {
+        String jarFilePath = jarFile.getPath();
+        ZipInputStream zipInput = null;
+        try {
+            ClasspathEntrySnapshotBuilder entryResourceCollectionBuilder = new ClasspathEntrySnapshotBuilder(stringInterner);
+            zipInput = new ZipInputStream(Files.newInputStream(Paths.get(jarFilePath)));
```

Comment:
Good catch. I fixed this.
Interestingly, `ZipInputStream` does not fail on an invalid zip file - for example a text file - but returns an empty set of entries. Should we try to detect this case? Or do you know a better implementation for treating zip files?